bluesky/tests/test_external_assets_and_paging.py

```diff
@@ -0,0 +1,433 @@
+from bluesky.utils import new_uid, IllegalMessageSequence
+from itertools import chain
+import pytest
+from bluesky.protocols import (
+    Asset, Readable,
+    Reading, SyncOrAsync, WritesExternalAssets, Flyable, EventCollectable, EventPageCollectable, Collectable
+)
+
+
+def read_Readable(self) -> Dict[str, Reading]:
+    return dict(x=dict(value=1.2, timestamp=0.0))
+
+
+def describe_Readable(self) -> Dict[str, DataKey]:
+    return dict(
+        y=dict(source="dummy", dtype="number", shape=[], external="STREAM:"),
+        x=dict(source="dummy", dtype="number", shape=[])
+    )
+
+
+def collect_asset_docs_Resource(self) -> Iterator[Asset]:
+    resource = PartialResource(
+        resource_kwargs={"some_argument": 1337},
+        root="awesome",
+        spec=".dat",
+        resource_path="no/really,/awesome",
+        uid=new_uid(),
+    )
+
+    return iter([("resource", resource)])
+
+
+def collect_asset_docs_Datum(self) -> Iterator[Asset]:
+    datum = Datum(
+        datum_id="some_resource_name/123123",
+        resource="uid_of_that_resource",
+        datum_kwargs={"some_argument": 1234},
+    )
+    return iter([("datum", datum)])
+
+
+def collect_asset_docs_StreamResource(self) -> Iterator[Asset]:
+    stream_resource = StreamResource(
+        resource_kwargs={"argument": 1},
+        resource_path="An/awesome/path",
+        root="some_detail",
+        spec=".hdf5",
+        uid=new_uid(),
+    )
+    return iter([("stream_resource", stream_resource)])
+
+
+def collect_asset_docs_StreamDatum(self) -> Iterator[Asset]:
+
+    stream_datum = StreamDatum(
+        block_idx=32,
+        stream_resource=new_uid(),
+        uid=new_uid(),
+        data_keys=["y"],
+        seq_nums={"a": 12},
+        indices={"b", 1},
+    )
+    return iter([("stream_datum", stream_datum)])
+
+
+@pytest.mark.parametrize(
+    'asset_type,collect_asset_docs_fun',
+    [
+        ("resource", collect_asset_docs_Resource),
+        ("datum", collect_asset_docs_Datum),
+        ("stream_resource", collect_asset_docs_StreamResource),
+        ("stream_datum", collect_asset_docs_StreamDatum),
+    ]
+)
+def test_rd_desc_different_asset_types(RE, asset_type, collect_asset_docs_fun):
+    class X(Readable, WritesExternalAssets):
+        read = read_Readable
+        describe = describe_Readable
+        collect_asset_docs = collect_asset_docs_fun
+        name = "x"
+
+    x = X()
+    collector = []
+    RE([
+            Msg("open_run", x),
+            Msg("create", name="x"),
+            Msg("read", x),
+            Msg("save", x),
+            Msg("close_run", x),
+        ],
+        lambda *args: collector.append(args)
+       )
+    assert len(collector) == 5
+    assert collector[2][0] == asset_type
+
+
+def describe_with_name(self) -> SyncOrAsync[Dict[str, DataKey]]:
+    data_keys = {
+        str(x): DataKey(shape=[1], source="a", dtype="string")
+        for x in ["x", "y", "z"]
```

Comment:
Please could we change these to be some more typical strings, like "det1" for datakeys, and "stream1" for stream names?